### PR TITLE
added dto, mappers, services, specification and controllers

### LINKED ISSUE #5: dto, mappers, services, specification and controllers
create added dto, mappers, services, specification and controllers

### PR DESCRIPTION
**Pull request by Bohdana-Zarembovska**
added dto, mappers, services, specification and controllers

### CODE REVIEW

File: src/main/java/org/electronicReferences/controllers/DiagnosisController.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    private final DiagnosisService diagnosisService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public DiagnosisDTO addDiagnosis(@Valid @RequestBody DiagnosisDTO diagnosisDTO){
+        return diagnosisService.addDiagnosis(diagnosisDTO);
+    }
+
+    @PutMapping("/{id}")
+    public DiagnosisDTO updateDiagnosis(@PathVariable Integer id, @RequestBody DiagnosisDTO diagnosisDTO) {
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/DiagnosisController.java]: You should use something like `DiagnosisUpdateDTO` with only the updatable fields

File: src/main/java/org/electronicReferences/controllers/DiagnosisController.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/diagnoses")
+public class DiagnosisController {
+    private final DiagnosisService diagnosisService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public DiagnosisDTO addDiagnosis(@Valid @RequestBody DiagnosisDTO diagnosisDTO){
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/DiagnosisController.java]: Same here but for create.

File: src/main/java/org/electronicReferences/controllers/DoctorController.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/doctors")
+public class DoctorController {
+    private final DoctorService doctorService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public DoctorDTO addDoctor(@Valid @RequestBody DoctorDTO doctorDTO){
+        return doctorService.addDoctor(doctorDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/DoctorController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/DoctorController.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @PutMapping("/{id}")
+    public DoctorDTO updateDoctor(@PathVariable Integer id, @RequestBody DoctorDTO doctorDTO) {
+        return doctorService.updateDoctor(id, doctorDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/DoctorController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/PatientController.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/patients")
+public class PatientController {
+    private final PatientService patientService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public PatientDTO addPatient(@Valid @RequestBody PatientDTO patientDTO){
+        return patientService.addPatient(patientDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/PatientController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/PatientController.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @PutMapping("/{id}")
+    public PatientDTO updatePatient(@PathVariable Integer id, @RequestBody PatientDTO patientDTO) {
+        return patientService.updatePatient(id, patientDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/PatientController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/ReferenceController.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/references")
+public class ReferenceController {
+    private final ReferenceService referenceService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ReferenceDTO addReference(@Valid @RequestBody ReferenceDTO referenceDTO){
+        return referenceService.addReference(referenceDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/ReferenceController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/ReferenceController.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @PutMapping("/{id}")
+    public ReferenceDTO updateReference(@PathVariable Integer id, @RequestBody ReferenceDTO referenceDTO) {
+        return referenceService.updateReference(id, referenceDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/ReferenceController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/UserController.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/users")
+public class UserController {
+    private final UserService userService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public UserDTO addUser(@Valid @RequestBody UserDTO userDTO){
+        return userService.addUser(userDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/UserController.java]: Same here

File: src/main/java/org/electronicReferences/controllers/UserController.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @PutMapping("/{id}")
+    public UserDTO updateUser(@PathVariable Integer id, @RequestBody UserDTO userDTO) {
+        return userService.updateUser(id, userDTO);
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/controllers/UserController.java]: Same here

File: src/main/java/org/electronicReferences/dto/UserDTO.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+public class UserDTO {
+    private Integer id;
+
+    @NotBlank(message = "Blood type must be specified")
+    private String fullName;
+
+    @NotBlank(message = "Date of birth must be specified")
+    private String phoneNumber;
+
+    @NotBlank(message = "Sex must be specified")
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/dto/UserDTO.java]: Change message

File: src/main/java/org/electronicReferences/dto/UserDTO.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Data;
+
+@Data
+public class UserDTO {
+    private Integer id;
+
+    @NotBlank(message = "Blood type must be specified")
+    private String fullName;
+
+    @NotBlank(message = "Date of birth must be specified")
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/dto/UserDTO.java]: Same

File: src/main/java/org/electronicReferences/dto/UserDTO.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
+package org.electronicReferences.dto;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class UserDTO {
+    private Integer id;
+
+    @NotBlank(message = "Blood type must be specified")
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/dto/UserDTO.java]: Same

File: src/main/java/org/electronicReferences/specification/UserSpecification.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package org.electronicReferences.specification;
+
+import org.electronicReferences.models.User;
+import org.springframework.stereotype.Component;
+import org.springframework.data.jpa.domain.Specification;
+
+@Component
+public class UserSpecification {
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/specification/UserSpecification.java:8]: Are you going to have multiple specifications?

**Review by justme1135 — CHANGES_REQUESTED** [review]

File: src/main/java/org/electronicReferences/controllers/DiagnosisController.java (hunk 1 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/DiagnosisController.java]: Done

File: src/main/java/org/electronicReferences/controllers/DiagnosisController.java (hunk 2 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/DiagnosisController.java]: Done

File: src/main/java/org/electronicReferences/controllers/DoctorController.java (hunk 3 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/DoctorController.java]: Done

File: src/main/java/org/electronicReferences/controllers/DoctorController.java (hunk 4 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/DoctorController.java]: Done

File: src/main/java/org/electronicReferences/controllers/PatientController.java (hunk 5 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/PatientController.java]: Done

File: src/main/java/org/electronicReferences/controllers/PatientController.java (hunk 6 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/PatientController.java]: Done

File: src/main/java/org/electronicReferences/controllers/ReferenceController.java (hunk 7 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/ReferenceController.java]: Done

File: src/main/java/org/electronicReferences/controllers/ReferenceController.java (hunk 8 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/ReferenceController.java]: Done

File: src/main/java/org/electronicReferences/controllers/UserController.java (hunk 9 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/UserController.java]: Done

File: src/main/java/org/electronicReferences/controllers/UserController.java (hunk 10 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/controllers/UserController.java]: Done

File: src/main/java/org/electronicReferences/dto/UserDTO.java (hunk 11 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/dto/UserDTO.java]: Done

File: src/main/java/org/electronicReferences/dto/UserDTO.java (hunk 12 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/dto/UserDTO.java]: Done

File: src/main/java/org/electronicReferences/dto/UserDTO.java (hunk 13 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/dto/UserDTO.java]: Done

File: src/main/java/org/electronicReferences/specification/UserSpecification.java (hunk 14 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/specification/UserSpecification.java:8]: No, only this one

File: src/main/java/org/electronicReferences/services/DiagnosisService.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+    public DiagnosisGetDTO addDiagnosis(DiagnosisCreateDTO diagnosisCreateDTO) {
+        Diagnosis diagnosis = diagnosisMapper.toEntity(diagnosisCreateDTO);
+        Diagnosis savedDiagnosis = diagnosisRepository.save(diagnosis);
+        return diagnosisMapper.toDTO(savedDiagnosis);
+    }
+
+    public DiagnosisGetDTO updateDiagnosis(Integer id, DiagnosisUpdateDTO diagnosisUpdateDTO) {
+        Diagnosis diagnosis = diagnosisRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("diagnosis with this id is not found"));
```
Comment (getmane) [inline src/main/java/org/electronicReferences/services/DiagnosisService.java]: move error message as a constant, also pass ID "this id" does not really help

File: src/main/java/org/electronicReferences/services/DoctorService.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+    public DoctorUpdateGetDTO addDoctor(DoctorCreateDTO doctorCreateDTO) {
+        Doctor doctor = doctorMapper.toEntity(doctorCreateDTO);
+        Doctor savedDoctor = doctorRepository.save(doctor);
+        return doctorMapper.toDTO(savedDoctor);
+    }
+
+    public DoctorUpdateGetDTO updateDoctor(Integer id, DoctorUpdateGetDTO doctorUpdateGetDTO) {
+        Doctor doctor = doctorRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("doctor with this id is not found"));
```
Comment (getmane) [inline src/main/java/org/electronicReferences/services/DoctorService.java]: Same here

File: src/main/java/org/electronicReferences/services/PatientService.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+    public PatientUpdateGetDTO addPatient(PatientCreateDTO patientCreateDTO) {
+        Patient patient = patientMapper.toEntity(patientCreateDTO);
+        Patient savedPatient = patientRepository.save(patient);
+        return patientMapper.toDTO(savedPatient);
+    }
+
+    public PatientUpdateGetDTO updatePatient(Integer id, PatientUpdateGetDTO patientUpdateGetDTO) {
+        Patient patient = patientRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("patient with this id is not found"));
```
Comment (getmane) [inline src/main/java/org/electronicReferences/services/PatientService.java]: And here

File: src/main/java/org/electronicReferences/services/ReferenceService.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+    public ReferenceUpdateGetDTO addReference(ReferenceCreateDTO referenceCreateDTO) {
+        Reference reference = referenceMapper.toEntity(referenceCreateDTO);
+        Reference savedReference = referenceRepository.save(reference);
+        return referenceMapper.toDTO(savedReference);
+    }
+
+    public ReferenceUpdateGetDTO updateReference(Integer id, ReferenceUpdateGetDTO referenceUpdateGetDTO) {
+        Reference reference = referenceRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("reference with this id is not found"));
```
Comment (getmane) [inline src/main/java/org/electronicReferences/services/ReferenceService.java]: And here

File: src/main/java/org/electronicReferences/services/UserService.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+    public UserUpdateGetDTO addUser(UserCreateDTO userCreateDTO) {
+        User user = userMapper.toEntity(userCreateDTO);
+        User savedUser = userRepository.save(user);
+        return userMapper.toDTO(savedUser);
+    }
+
+    public UserUpdateGetDTO updateUser(Integer id, UserUpdateGetDTO userUpdateGetDTO) {
+        User user = userRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("user with this id is not found"));
```
Comment (getmane) [inline src/main/java/org/electronicReferences/services/UserService.java]: And here

File: src/main/resources/application.properties [hunk 20]
```diff
@@ -3,4 +3,5 @@ spring.datasource.url=jdbc:mysql://${MYSQL_HOST:localhost}:3306/electronicrefere
 spring.datasource.username=root
 spring.datasource.password=7CiUaX359m
```
Comment (getmane) [inline src/main/resources/application.properties]: remove your password (replace with "root", "password" or smth like that)

File: src/main/java/org/electronicReferences/dto/DoctorDTOs/DoctorCreateDTO.java [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+
+@Data
+public class DoctorCreateDTO {
+    private Integer id;
+
+    @NotBlank(message = "Specification type must be specified")
+    private String specification;
+
+    @NotBlank(message = "Hospital name of birth must be specified")
```
Comment (justme1135) [inline src/main/java/org/electronicReferences/dto/DoctorDTOs/DoctorCreateDTO.java]: `"Hospital name must be specified"`?

**Review by justme1135 — APPROVED** [review]
One more thing

File: src/main/java/org/electronicReferences/services/DiagnosisService.java (hunk 15 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/services/DiagnosisService.java]: Done

File: src/main/java/org/electronicReferences/services/DoctorService.java (hunk 16 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/services/DoctorService.java]: Done

File: src/main/java/org/electronicReferences/services/PatientService.java (hunk 17 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/services/PatientService.java]: Done

File: src/main/java/org/electronicReferences/services/ReferenceService.java (hunk 18 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/services/ReferenceService.java]: Done

File: src/main/java/org/electronicReferences/services/UserService.java (hunk 19 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/services/UserService.java]: Done

File: src/main/resources/application.properties (hunk 20 above)
Comment (Bohdana-Zarembovska) [reply src/main/resources/application.properties]: Done

File: src/main/java/org/electronicReferences/dto/DoctorDTOs/DoctorCreateDTO.java (hunk 21 above)
Comment (Bohdana-Zarembovska) [reply src/main/java/org/electronicReferences/dto/DoctorDTOs/DoctorCreateDTO.java]: Done

**Review by getmane — APPROVED** [review]

**Pull request merged**